### PR TITLE
Blog tags

### PR DESCRIPTION
**Pull request by E-zh**

### CODE REVIEW

File: src/Http/Controllers/Admin/BlogPostController.php [hunk 1]
```diff
@@ -19,7 +21,7 @@ public function index()
         \Title::prepend(trans('dashboard.title.prepend'));
         \Title::append(trans('blog::blog.post.title'));
 
-        $posts = Post::paginate(20);
+        $posts = Post::orderBy('id', 'desc')->paginate(20);
```
Comment (garf) [inline src/Http/Controllers/Admin/BlogPostController.php]: Better order by `published_at`

File: src/Http/Controllers/Admin/BlogPostController.php [hunk 2]
```diff
@@ -38,12 +40,27 @@ public function edit($id, Post $post)
     {
         $post = $post->find($id);
 
+        // It's first version of code. It's we need?
```
Comment (garf) [inline src/Http/Controllers/Admin/BlogPostController.php]: https://laracasts.com/discuss/channels/eloquent/fetch-most-used-tags
Check this for most used tags fetching

File: src/Http/Controllers/Admin/BlogPostController.php [hunk 3]
```diff
@@ -91,4 +110,29 @@ public function destroy($id)
...
     }
+
+    private function _setTags($tags_str, $post_id)
+    {
+        PostTags::where('post_id', $post_id)->delete();
+
+        $tags = explode(',', $tags_str);
+
+        foreach ($tags as $tag) {
+            if (trim($tag) == '') {
```
Comment (garf) [inline src/Http/Controllers/Admin/BlogPostController.php]: You trimming them twice. Let's make it DRY :)

File: src/Http/Controllers/Admin/BlogPostController.php [hunk 4]
```diff
@@ -91,4 +110,29 @@ public function destroy($id)
...
+                continue;
+            }
+            $tag = mb_strtolower(trim($tag));
+            $dbtag = Tags::where('tag', 'like', $tag)->first();
+            if (empty($dbtag)) {
+                $dbtag = new Tags();
+                $dbtag->tag = strip_tags($tag);
+                $dbtag->save();
+            }
+            $post_tag = new PostTags();
```
Comment (garf) [inline src/Http/Controllers/Admin/BlogPostController.php]: Why not using eloquent relation methods? 
Check this: https://laravel.com/docs/5.3/eloquent-relationships#inserting-and-updating-related-models

File: src/Http/Controllers/BlogController.php [hunk 5]
```diff
@@ -40,4 +41,19 @@ public function post(Post $post, $slug)
 
         return view('blog::site.post', compact('post'));
     }
+
+    public function getTag(Tags $tags, $tag)
```
Comment (garf) [inline src/Http/Controllers/BlogController.php]: `getByTag`

File: src/Http/Controllers/BlogController.php [hunk 6]
```diff
@@ -40,4 +41,19 @@ public function post(Post $post, $slug)
...
+    {
+        $tags = $tags->where('tag', $tag)->first();
+
+        if(!isset($tags)){
```
Comment (garf) [inline src/Http/Controllers/BlogController.php:49]: No need to this checks. Just select posts by tag. And if no posts, show that no posts. but not 404.

File: src/Models/Tags.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    public function posts()
+    {
+        return $this->belongsToMany('Ourgarage\Blog\Models\Post', 'post_tags', 'tag_id', 'post_id');
+    }
+
+    /**
+     * @param integer $take
+     * @return object
+     */
+    public static function popularTags($take)
```
Comment (garf) [inline src/Models/Tags.php]: What is this? :) Why it's static? :) Use it in normal way please.

File: src/database/migrations/2016_11_18_044430_create_post_tags_table.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('post_tags', function (Blueprint $table) {
+            $table->increments('id');
+            $table->integer('post_id');
+            $table->integer('tag_id');
+            $table->timestamps();
```
Comment (garf) [inline src/database/migrations/2016_11_18_044430_create_post_tags_table.php]: I think we don't need timestamps here.

File: src/database/migrations/2016_11_18_044430_create_post_tags_table.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('post_tags');
```
Comment (garf) [inline src/database/migrations/2016_11_18_044430_create_post_tags_table.php]: don't use `dropIfExists()`. Use `drop()` It should exist if migrations runned in correct way. Error will indicate us that something goes wrong with data structure.

File: src/database/migrations/2016_11_18_043230_create_tags_table.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('tags');
```
Comment (garf) [inline src/database/migrations/2016_11_18_043230_create_tags_table.php]: don't use `dropIfExists()`. Use `drop()`

File: src/resources/assets/js/post.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+$('.add-tag').on('click', function(){
```
Comment (garf) [inline src/resources/assets/js/post.js]: space before `{`

File: src/resources/assets/js/post.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    var inputTags = $('#inputTags');
```
Comment (garf) [inline src/resources/assets/js/post.js]: prefix selectors with `$`
`$inputTags`

File: src/resources/assets/js/post.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+    var newTag = $(this).attr('data-tag');
+    $('#inputTags').tagsinput('add', newTag);
```
Comment (garf) [inline src/resources/assets/js/post.js]: Use already created selector $inputTags.

File: src/resources/assets/js/post.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+});
+
+var param = $('#date_published').data('datetimepicker-locale');
```
Comment (garf) [inline src/resources/assets/js/post.js]: Why you keep in `data-datetimepicker-locale` something else than locale?

File: src/resources/views/admin/post/post.blade.php [hunk 15]
```diff
@@ -98,21 +98,42 @@
...
+                                </div>
+                            </div>
+                        @endif
+
+                        <div class="form-group has-feedback">
+                            <label class="col-md-2">{{ trans('blog::blog.post.date-published') }} :</label>
+                            <div class="col-md-2">
+                                <input id="date_published" type="text" name="date_published"
+                                       class="form-control"
+                                       data-datetimepicker-locale='{"lang": "{{ app()->getLocale() }}",
```
Comment (garf) [inline src/resources/views/admin/post/post.blade.php]: Here. In `data-datetimepicker-locale` should be only locale info.
```
data-datetimepicker-locale="{{ app()->getLocale() }}"
data-datetimepicker-format="{{ trans('date.format.full.js') }}"

**Review by garf — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/BlogController.php (hunk 6 above)
Comment (E-zh) [reply src/Http/Controllers/BlogController.php:49]: This check is not post. If the tag is entered in the URL is not - we have to get 404. Or not?

Comment (garf) [reply src/Http/Controllers/BlogController.php:49]: No. This is not the page with one record. We have a list of records here. So if no records, we show it. But not 404.

File: src/Http/Controllers/Admin/BlogPostController.php [hunk 16]
```diff
@@ -34,16 +36,18 @@ public function add()
         return view('blog::admin.post.post', compact('categories'));
     }
 
-    public function edit($id, Post $post)
+    public function edit(Tags $tags, $id, Post $post)
```
Comment (garf) [inline src/Http/Controllers/Admin/BlogPostController.php]: Order of params should be adjusted. :)

**Review by garf — APPROVED** [review]

**Pull request merged**